### PR TITLE
Migrate time sync test cases.

### PR DESCRIPTION
**Pull request by LiliDeng**

### CODE REVIEW

File: microsoft/testsuites/core/timesync.py [hunk 1]
```diff
@@ -64,3 +68,45 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+        """,
+        priority=2,
+    )
+    def timesync_check_timesource(self, environment: Environment, node: Node) -> None:
+        # 1. Check clock source name shown as hyperv_clocksource_tsc_page
+        clocksource = "hyperv_clocksource_tsc_page"
+        cat = node.tools[Cat]
+        clock_source_result = cat.run(self.current_clocksource)
+        os_version = node.os.os_version
+        os_verion_convert = (
```
Comment (glima) [inline microsoft/testsuites/core/timesync.py]: Typo?

File: microsoft/testsuites/core/timesync.py [hunk 2]
```diff
@@ -64,3 +68,45 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+            os_version.release.split(".")[0] + "." + os_version.release.split(".")[1]
```
Comment (glima) [inline microsoft/testsuites/core/timesync.py]: Maybe .rpartition('.')[0] is simpler?

File: lisa/operating_system.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ class OperatingSystem:
     )
     __os_release_pattern_id = re.compile(r"^ID=\"?([^\" \r\n]+)[^\" \n]*\"?\r?$", re.M)
     __redhat_release_pattern_header = re.compile(r"^([^ ]*) .*$")
-    __redhat_release_pattern_bracket = re.compile(r"^.*\(([^ ]*).*\)$")
+    redhat_release_pattern_bracket = re.compile(r"^.*\(([^ ]*).*\)$")
```
Comment (squirrelsc) [inline lisa/operating_system.py]: If it's changed to public, move it above other private members.

File: lisa/operating_system.py [hunk 4]
```diff
@@ -409,7 +409,7 @@ class OpenBSD(BSD):
 
 
 class Fedora(Linux):
-    __fedora_release_pattern_version = re.compile(r"^.*release\s+([0-9\.]+).*$")
+    fedora_release_pattern_version = re.compile(r"^.*release\s+([0-9\.]+).*$")
```
Comment (squirrelsc) [inline lisa/operating_system.py]: It looks don't need to be public.

File: lisa/operating_system.py (hunk 3 above)
Comment (squirrelsc) [reply lisa/operating_system.py]: Let's move the shared pattern out of the base class, put in into the file level, and keep it like `_redhat_release_pattern_bracket`

File: lisa/tools/lscpu.py [hunk 5]
```diff
@@ -11,9 +11,11 @@
 
 class Lscpu(Tool):
     __vcpu_sockets = re.compile(r"^CPU\(s\):[ ]+([\d]+)\r?$", re.M)
+    __architecture = re.compile(r"^Architecture:\s+(.*)?\r$", re.M)
```
Comment (squirrelsc) [inline lisa/tools/lscpu.py]: give an example of output.

File: lisa/tools/lscpu.py [hunk 6]
```diff
@@ -26,6 +28,17 @@ def _windows_tool(cls) -> Optional[Type[Tool]]:
     def _check_exists(self) -> bool:
         return True
 
+    def get_architecture(self, force_run: bool = False) -> str:
+        result = self.run(force_run=force_run)
+        matched = self.__architecture.findall(result.stdout)
```
Comment (squirrelsc) [inline lisa/tools/lscpu.py]: The name `__architecture` is confusing with `_architecture`. So rename the `__architecture` like `__architecture_pattern`

File: lisa/tools/lscpu.py [hunk 7]
```diff
@@ -26,6 +28,17 @@ def _windows_tool(cls) -> Optional[Type[Tool]]:
...
+        assert_that(
+            len(matched),
+            f"architecture should have exact one line, but got {matched}",
+        ).is_equal_to(1)
+        self._architecture = matched[0]
+
+        return self._architecture
```
Comment (squirrelsc) [inline lisa/tools/lscpu.py]: `self._architecture` doesn't need be an instance member here, because it's not cached, or reused. Use a local variable.

File: microsoft/testsuites/core/timesync.py [hunk 8]
```diff
@@ -64,3 +68,45 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+            elif float(os_verion_convert) <= 7.2:
+                # 1.2 If it is centos/rhel/oracle below 7.2 + without LIS driver, it
+                # shown as hyperv_clocksource.
+                clocksource = "hyperv_clocksource"
+        assert_that(clock_source_result.stdout).is_equal_to(clocksource)
+        # 2. Check CPU flag contains tsc from /proc/cpuinfo.
+        lscpu = node.tools[Lscpu]
+        if "x86_64" == lscpu.get_architecture():
+            cpu_info_result = cat.run("/proc/cpuinfo")
+            assert_that(cpu_info_result.stdout).contains("tsc")
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: add description for each assert.

File: lisa/operating_system.py (hunk 4 above)
Comment (LiliDeng) [reply lisa/operating_system.py]: I can't use it from RedHat class if it is not public.

Comment (sharsonia) [reply lisa/operating_system.py]: Redhat inherits Fedora class, so should be accessible from there like this - super().__fedora_release_pattern_version

Comment (LiliDeng) [reply lisa/operating_system.py]: I tried but didn't work, let me try again.

File: microsoft/testsuites/core/timesync.py [hunk 9]
```diff
@@ -64,3 +68,45 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+        # 1. Check clock source name shown as hyperv_clocksource_tsc_page
+        clocksource = "hyperv_clocksource_tsc_page"
+        cat = node.tools[Cat]
+        clock_source_result = cat.run(self.current_clocksource)
+        os_version = node.os.os_version
+        os_verion_convert = (
+            os_version.release.split(".")[0] + "." + os_version.release.split(".")[1]
+        )
+        if isinstance(node.os, Redhat):
+            if node.os._package_exists("kmod-microsoft"):
```
Comment (sharsonia) [inline microsoft/testsuites/core/timesync.py]: I have a commit which is adding LIS driver as a tool and can get_version() to check if LIS drivers are installed or not. May be we can use that tool if that commit gets merged.

File: microsoft/testsuites/core/timesync.py [hunk 10]
```diff
@@ -64,3 +68,45 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+                if float(os_verion_convert) < 7.5:
```
Comment (sharsonia) [inline microsoft/testsuites/core/timesync.py]: you should be able to directly compare it like 
if float(self.node.os.os_version.release) < 7.5:

File: lisa/operating_system.py (hunk 4 above)
Comment (squirrelsc) [reply lisa/operating_system.py]: `__` is private, cannot be used in subclasses, `_` can be.

File: microsoft/testsuites/core/timesync.py (hunk 10 above)
Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: @sharsonia  Can we provide a method in os to compare the os_version? For most versions, they can be compared like float, some of them may need be customized, for example "12.3.5". But it looks doable.

File: microsoft/testsuites/core/timesync.py (hunk 9 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: sure

File: microsoft/testsuites/core/timesync.py (hunk 10 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: yeah, I used float(self.node.os.os_version.release), but failed when the release returns 8.3.2011

File: microsoft/testsuites/core/timesync.py (hunk 9 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: I found the lisdirver tool is not suitable here, once this tool is invoked, it will try to install it.

Comment (sharsonia) [reply microsoft/testsuites/core/timesync.py]: Hmm we can may be add a line to check_exist() before installing it in the _install() method.
Tis will let it not install if already installed.

Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: current logic is like this, when not exist, it will install it, but in my case, I just want to judge it exists or not, no need to install it.

File: lisa/operating_system.py [hunk 11]
```diff
@@ -17,6 +17,7 @@
 
 
 _get_init_logger = partial(get_logger, name="os")
+_redhat_release_pattern_bracket = re.compile(r"^.*\(([^ ]*).*\)$")
```
Comment (squirrelsc) [inline lisa/operating_system.py:21]: Add examples above.

File: lisa/operating_system.py [hunk 12]
```diff
@@ -372,7 +372,9 @@ def _package_exists(self, package: str, signed: bool = True) -> bool:
 
     def _get_os_version(self) -> OsVersion:
         os_version = OsVersion("")
-        cmd_result = self._node.execute(cmd="lsb_release -as", no_error_log=True)
+        cmd_result = self._node.execute(
+            cmd="lsb_release -ar 2>/dev/null", shell=True, no_error_log=True
```
Comment (squirrelsc) [inline lisa/operating_system.py]: Add reason here, why it needs to redirect errors to null. If there is errors, how we know the reason?

File: lisa/sut_orchestrator/azure/tools.py [hunk 13]
```diff
@@ -62,22 +62,27 @@ def get_generation(self) -> str:
...
     def dependencies(self) -> List[Type[Tool]]:
         return [Wget, Modinfo]
 
     @property
     def can_install(self) -> bool:
-        if (
-            isinstance(self.node.os, Redhat)
-            and float(self.node.os.os_version.release) < 7.8
-        ):
+        if len(self.node.os.os_version.release.split(".")) > 2:
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/tools.py]: Add comments here to explain the logic. BTW, I think we can split versions to an int/str array, and use a method to compare it. For example, 8.3 can be [8, 3], and 8.3.20210609 can be [8, 3, 20210609]. They are easy to compare within same format. Anyway, you can add a task to me to support it.

File: lisa/tools/lscpu.py [hunk 14]
```diff
@@ -26,6 +29,18 @@ def _windows_tool(cls) -> Optional[Type[Tool]]:
...
         return True
 
+    def get_architecture(self, force_run: bool = False) -> str:
+        architecture: str = ""
+        result = self.run(force_run=force_run)
+        matched = self.__architecture_pattern.findall(result.stdout)
+        assert_that(
+            len(matched),
+            f"architecture should have exact one line, but got {matched}",
+        ).is_equal_to(1)
```
Comment (squirrelsc) [inline lisa/tools/lscpu.py]: use `is_length`, so it can output the original value, if it's wrong.

File: lisa/tools/lscpu.py [hunk 15]
```diff
@@ -26,6 +29,18 @@ def _windows_tool(cls) -> Optional[Type[Tool]]:
...
+        architecture = matched[0]
```
Comment (squirrelsc) [inline lisa/tools/lscpu.py:47]: We can have a validation, if it's a valid architecture string.

File: lisa/util/process.py [hunk 16]
```diff
@@ -77,13 +77,13 @@ def start(
             if not self._is_posix:
                 split_command = ["cmd", "/c", command]
             elif sudo:
-                command = f"sudo -s {command}"
+                command = f"sudo -s sh -c '{command}'"
```
Comment (squirrelsc) [inline lisa/util/process.py]: Why change this? It may have compatibility issues.

Comment (squirrelsc) [reply lisa/util/process.py]: `sudo -s` means sudo with shell already.

File: lisa/util/process.py [hunk 17]
```diff
@@ -77,13 +77,13 @@ def start(
...
                 split_command = shlex.split(command, posix=self._is_posix)
             else:
                 split_command = ["sh", "-c", command]
         else:
             if sudo and self._is_posix:
-                command = f"sudo {command}"
+                command = f"sudo sh -c '{command}'"
```
Comment (squirrelsc) [inline lisa/util/process.py]: this means sudo without shell. If add it, there is no way to disable shell with sudo.

File: microsoft/testsuites/core/timesync.py [hunk 18]
```diff
@@ -17,6 +18,13 @@ class TimeSync(TestSuite):
     ptp_registered_msg = "PTP clock support registered"
     hyperv_ptp_udev_rule = "ptp_hyperv"
     chrony_path = ["/etc/chrony.conf", "/etc/chrony/chrony.conf"]
+    current_clocksource = (
+        "/sys/devices/system/clocksource/clocksource0/current_clocksource"
+    )
+    available_clocksource = (
+        "/sys/devices/system/clocksource/clocksource0/available_clocksource"
+    )
+    unbind_file = "/sys/devices/system/clocksource/clocksource0/unbind_clocksource"
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: as above names, how about call it `unbind_clocksource`?

File: microsoft/testsuites/core/timesync.py [hunk 19]
```diff
@@ -64,3 +72,76 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+        ).is_subset_of(clocksource)
+
+        # 2. Check CPU flag contains tsc from /proc/cpuinfo.
+        lscpu = node.tools[Lscpu]
+        if "x86_64" == lscpu.get_architecture():
+            cpu_info_result = cat.run("/proc/cpuinfo")
+            assert_that(cpu_info_result.stdout).described_as(
+                "Expected 'constant_tsc' shown up in cpu flags."
+            ).contains("constant_tsc")
+        # 3. Check clocksource name shown up in dmesg.
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py:138]: add a blank line between each check.

File: microsoft/testsuites/core/timesync.py [hunk 20]
```diff
@@ -64,3 +72,76 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+        dmesg = node.tools[Dmesg]
+        assert_that(dmesg.get_output()).described_as(
+            f"Expected clocksource {clock_source_result.stdout} shown up in dmesg."
+        ).contains(f"clocksource {clock_source_result.stdout}")
+        # 4. Unbind current clock source if there are 2+ clock sources,
+        # check current clock source switch to a different one.
+        if node.shell.exists(PurePosixPath(f"{self.unbind_file}")):
+            available_clocksources = cat.run(self.available_clocksource)
+            available_clocksources_array = available_clocksources.stdout.split(" ")
+            if len(available_clocksources_array) >= 2:
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: explain the logic, why need it.

File: lisa/sut_orchestrator/azure/tools.py (hunk 13 above)
Comment (sharsonia) [reply lisa/sut_orchestrator/azure/tools.py:81]: Since the old logic was only considering version if isinstance(self.node.os, Redhat) , I think we don't need to worry about the version with format *.*.*

Comment (sharsonia) [reply lisa/sut_orchestrator/azure/tools.py]: > Add comments here to explain the logic. BTW, I think we can split versions to an int/str array, and use a method to compare it. For example, 8.3 can be [8, 3], and 8.3.20210609 can be [8, 3, 20210609]. They are easy to compare within same format. Anyway, you can add a task to me to support it.

Despite that we don't need to worry about the os version format issue here, I agree we can make version like you suggested. I will push out a PR for this today.

Comment (sharsonia) [inline lisa/sut_orchestrator/azure/tools.py]: Since the old logic was only considering version if isinstance(self.node.os, Redhat) , I think we don't need to worry about the version with format   like 8.3.20210609

Comment (squirrelsc) [reply lisa/sut_orchestrator/azure/tools.py]: It can be a Version class, it supports `<`, `==`, `>`, `raw` methods.

Comment (glima) [reply lisa/sut_orchestrator/azure/tools.py]: Let's not re-invent the wheel: https://pypi.org/project/semver/

File: microsoft/testsuites/core/timesync.py [hunk 21]
```diff
@@ -64,3 +72,76 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+                1.2 If it is centos/rhel/oracle below 7.2 + without LIS driver, it
+                    shown as hyperv_clocksource.
+            2. Check CPU flag contains tsc from /proc/cpuinfo.
+            3. Check clocksource name shown up in dmesg.
+            4. Unbind current clock source if there are 2+ clock sources, check current
+             clock source switch to a different one.
+        """,
+        priority=1,
+    )
+    def timesync_check_clocksource(self, environment: Environment, node: Node) -> None:
```
Comment (sharsonia) [inline microsoft/testsuites/core/timesync.py]: Do we need environment here?

File: lisa/sut_orchestrator/azure/tools.py [hunk 22]
```diff
@@ -62,22 +62,27 @@ def get_generation(self) -> str:
 
 class LisDriver(Tool):
     """
-    This is a virtual tool to detect/install LIS (Linus Integration Services) drivers.
+    This is a virtual tool to detect/install LIS (Linux Integration Services) drivers.
```
Comment (sharsonia) [inline lisa/sut_orchestrator/azure/tools.py]: Can we put the changes in this file in a separate commit please?

File: lisa/sut_orchestrator/azure/tools.py (hunk 13 above)
Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/tools.py]: why? 
I will revert changes for this LisDriver, since you will have a PR soon.

File: lisa/sut_orchestrator/azure/tools.py (hunk 22 above)
Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/tools.py]: sure, I will not contain this change, if you can submit one for this, that would be great.

File: lisa/sut_orchestrator/azure/tools.py (hunk 13 above)
Comment (sharsonia) [reply lisa/sut_orchestrator/azure/tools.py]: Because os_version for Redhat or CentOs, we are getting from cat /etc/redhat-release. I think we will get those only with major and minor version

cat /etc/redhat-release
Red Hat Enterprise Linux Server release 6.7 (Santiago)

But yeah sure you can revert all the changes. if you want to compare versions, you can just use what Gaustavo suggested.

Comment (LiliDeng) [reply lisa/sut_orchestrator/azure/tools.py]: I got exception when float(release) + image OpenLogic CentOS 8_3 latest, at that time the release is 8.3.2011, that's why I made this change.
cat /etc/redhat-release
CentOS Linux release 8.3.2011

Comment (sharsonia) [reply lisa/sut_orchestrator/azure/tools.py]: Got it. I see, yeah can compare with semver.compare(). you can make the change if you want but keep it in a different commit or else, I will push out a new PR. Either way is fine.

Comment (squirrelsc) [reply lisa/sut_orchestrator/azure/tools.py]: I'm doubt if all Linux distros follow sematic versions. If we use semver, make sure it fits to all versions. Another way is to implement different comparable version for different distros.

File: lisa/operating_system.py [hunk 23]
```diff
@@ -372,7 +373,16 @@ def _package_exists(self, package: str, signed: bool = True) -> bool:
 
     def _get_os_version(self) -> OsVersion:
         os_version = OsVersion("")
-        cmd_result = self._node.execute(cmd="lsb_release -as", no_error_log=True)
+        # Output of 'lsb_release -as' doesn't match the pattern.
```
Comment (squirrelsc) [inline lisa/operating_system.py]: This comments explained the history of change, it should be in commit message, not in code comments. The code comments should show examples of outputs, which distro it's from, and which one is tricky.

File: lisa/operating_system.py [hunk 24]
```diff
@@ -409,7 +419,7 @@ class OpenBSD(BSD):
 
 
 class Fedora(Linux):
-    __fedora_release_pattern_version = re.compile(r"^.*release\s+([0-9\.]+).*$")
+    _fedora_release_pattern_version = re.compile(r"^.*release\s+([0-9\.]+).*$")
```
Comment (squirrelsc) [inline lisa/operating_system.py:419]: Add examples

File: microsoft/testsuites/core/timesync.py (hunk 20 above)
Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: ok, so `> 1` is easier to understand than `>= 2` here.

File: microsoft/testsuites/core/timesync.py [hunk 25]
```diff
@@ -64,3 +74,78 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+                timeout = 30
+                while timout_timer.elapsed(False) < timeout:
+                    current_clock_source_result = cat.run(
+                        self.current_clocksource, force_run=True
+                    )
+                    if (
+                        current_clock_source_result.stdout
+                        in available_clocksources_array
+                    ):
+                        break
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: Add a sleep here to prevent access too fast locally, and slow down other things. May be 0.5 seconds.

File: lisa/operating_system.py [hunk 26]
```diff
@@ -372,7 +373,16 @@ def _package_exists(self, package: str, signed: bool = True) -> bool:
...
         os_version = OsVersion("")
-        cmd_result = self._node.execute(cmd="lsb_release -as", no_error_log=True)
+        # Output of 'lsb_release -as' doesn't match the pattern.
+        # No LSB modules are available.
+        # Debian
+        # Debian GNU/Linux 10 (buster)
+        # 10
+        # buster
+        # Change it into 'lsb_release -ar'.
+        cmd_result = self._node.execute(
```
Comment (sharsonia) [inline lisa/operating_system.py:379]: This is what I see from the man page - 

-v, --version
Show the version of the LSB against which your current installation is compliant. The version is expressed as a colon separated list of LSB module descriptions.
-i, --id
Display the distributor's ID.
-d, --description
Display a description of the currently installed distribution.
-r, --release
Display the release number of the currently installed distribution.
-c, --codename
Display the code name of the currently installed distribution.
-a, --all
Display all of the above information.
-s, --short
Use the short output format for any information displayed. This format omits the leading header(s).

I think we just need lsb_release -a
We don't even need option "r" because if we are using option "a" then option "r" is not useful.

~$ lsb_release -as
No LSB modules are available.
Ubuntu
Ubuntu 20.04.2 LTS
20.04
focal

~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal

~$ lsb_release -ar
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal

Comment (LiliDeng) [reply lisa/operating_system.py:379]: yeah, I will use -a only

File: lisa/util/process.py [hunk 27]
```diff
@@ -77,7 +77,7 @@ def start(
             if not self._is_posix:
                 split_command = ["cmd", "/c", command]
             elif sudo:
-                command = f"sudo -s {command}"
+                command = f"sudo -s sh -c '{command}'"
```
Comment (squirrelsc) [inline lisa/util/process.py]: Try to construct a string list like below, not rely on shlex.
`split_command = ["sudo", "sh", "-c", command]`

File: microsoft/testsuites/core/timesync.py (hunk 20 above)
Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: not see the second is addressed

File: microsoft/testsuites/core/timesync.py [hunk 28]
```diff
@@ -64,3 +75,79 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+                    sudo=True,
+                )
+                assert_that(cmd_result.exit_code).described_as(
+                    f"Fail to execute command "
+                    f"[echo {clock_source_result.stdout} > {self.unbind_clocksource}]."
+                ).is_equal_to(0)
+                timout_timer = create_timer()
+                timeout = 30
+                while timout_timer.elapsed(False) < timeout:
+                    time.sleep(5)
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: 5 is too large, try 0.5. And don't put it at beginning of the loop, put it at the end. If it's very fast, it may not to wait.

File: microsoft/testsuites/core/timesync.py [hunk 29]
```diff
@@ -64,3 +75,79 @@ def timesync_validate_ptp(self, environment: Environment, node: Node) -> None:
...
+        # 4. Unbind current clock source if there are 2+ clock sources,
+        # check current clock source can be switched to a different one.
+        if node.shell.exists(PurePosixPath(f"{self.unbind_clocksource}")):
+            available_clocksources = cat.run(self.available_clocksource)
+            available_clocksources_array = available_clocksources.stdout.split(" ")
+            # We can not unbind clock source if there is only one existed.
+            if len(available_clocksources_array) >= 2:
+                available_clocksources_array.remove(clock_source_result.stdout)
+                cmd_result = node.execute(
+                    f"sh -c 'echo {clock_source_result.stdout} > "
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: use shell directly? Don't use it here.

File: lisa/tools/lscpu.py (hunk 15 above)
Comment (squirrelsc) [reply lisa/tools/lscpu.py:47]: where is the validation? Check if it's `x86_64` or other valid strings.

Comment (LiliDeng) [reply lisa/tools/lscpu.py:47]: need figure out what are the valid values.

Comment (squirrelsc) [reply lisa/tools/lscpu.py:47]: we can add it one by one. If it parses incorrect value, the error is easier to find.

File: lisa/util/process.py [hunk 30]
```diff
@@ -77,8 +77,7 @@ def start(
             if not self._is_posix:
                 split_command = ["cmd", "/c", command]
             elif sudo:
-                command = f"sudo -s {command}"
-                split_command = shlex.split(command, posix=self._is_posix)
+                split_command = ["sudo", "-s", "sh", "-c", command]
```
Comment (squirrelsc) [inline lisa/util/process.py]: "-s" is not necessary now.

File: lisa/tools/lscpu.py [hunk 31]
```diff
@@ -57,6 +57,10 @@ def get_core_count(self, force_run: bool = False) -> int:
 
         return self._core_count
 
+    def is_amd_cpu(self, force_run: bool = False) -> bool:
```
Comment (squirrelsc) [inline lisa/tools/lscpu.py]: It's better to return an Enum like CpuType.

File: microsoft/testsuites/core/timesync.py [hunk 32]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+            clock_event_times = timer_list_result.stdout.count(f"{clock_event_name}")
+            assert_that(clock_event_times).described_as(
+                f"Expected {clock_event_name} shown up {core_count} times in output "
+                f"of /proc/timer_list, but actual it shows up "
+                f"{clock_event_times} times."
+            ).is_equal_to(core_count)
+
+            # 3. when cpu count is 1 and cpu type is not AMD type, unbind current time
+            #  clock event, check current time clock event switch to 'lapic'.
+            if not lscpu.is_amd_cpu() and 1 == core_count:
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: Does it applies to Intel CPU? if so, check it's with a positive condition, not negative.

File: microsoft/testsuites/core/timesync.py [hunk 33]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+    @TestCaseMetadata(
+        description="""
+        This test is to check -
+            1. Current clock event name is 'Hyper-V clockevent'.
+            2. 'Hyper-V clockevent' and 'hrtimer_interrupt' show up times in
+             /proc/timer_list should equal to cpu count.
+            3. when cpu count is 1 and cpu type is not AMD type, unbind current time
+             clock event, check current time clock event switch to 'lapic'.
+        """,
+        priority=1,
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: Does it need to a p1 test case? if it's failed, what's the impact?

File: microsoft/testsuites/core/timesync.py [hunk 34]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
                     f"in {available_clocksources_array}."
                 ).contains(current_clock_source_result.stdout)
+
+    @TestCaseMetadata(
+        description="""
+        This test is to check -
+            1. Current clock event name is 'Hyper-V clockevent'.
+            2. 'Hyper-V clockevent' and 'hrtimer_interrupt' show up times in
+             /proc/timer_list should equal to cpu count.
+            3. when cpu count is 1 and cpu type is not AMD type, unbind current time
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: Is there any document about this behavior?

File: microsoft/testsuites/core/timesync.py [hunk 35]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+             clock event, check current time clock event switch to 'lapic'.
+        """,
+        priority=1,
+    )
+    def timesync_check_unbind_clockevent(self, node: Node) -> None:
+        if node.shell.exists(PurePosixPath(f"{self.current_clockevent}")):
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: It's a string, not need to wrap it by quote.
`PurePosixPath(self.current_clockevent)`

File: microsoft/testsuites/core/timesync.py [hunk 36]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+            # 1. Current clock event name is 'Hyper-V clockevent'.
+            clock_event_name = "Hyper-V clockevent"
+            event_handler_name = "hrtimer_interrupt"
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: this value is used in next step, so move it to next step is clearer.

File: microsoft/testsuites/core/timesync.py [hunk 37]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+            cat = node.tools[Cat]
+            clock_event_result = cat.run(self.current_clockevent)
+            assert_that(clock_event_result.stdout).described_as(
+                f"Expected clockevent name is {clock_event_name}, "
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py:188]: The actual and expected will be printed by assertpy. The description should explain why this needed, how to fix or the reference.

File: microsoft/testsuites/core/timesync.py [hunk 38]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+            # 2. 'Hyper-V clockevent' and 'hrtimer_interrupt' show up times in
+            #  /proc/timer_list should equal to cpu count.
+            timer_list_result = cat.run("/proc/timer_list", sudo=True)
+            lscpu = node.tools[Lscpu]
+            core_count = lscpu.get_core_count()
+            event_handler_times = timer_list_result.stdout.count(
+                f"{event_handler_name}"
+            )
+            assert_that(event_handler_times).described_as(
+                f"Expected {event_handler_name} shown up {core_count} times in output "
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py:202]: check what assertpy can show, and make sure there is no duplicate information, and easy to understand. The expected value and actual value should be printed already.

File: microsoft/testsuites/core/timesync.py [hunk 39]
```diff
@@ -151,3 +153,74 @@ def timesync_check_clocksource(self, node: Node) -> None:
...
+                    sudo=True,
+                    shell=True,
+                )
+                assert_that(cmd_result.exit_code).described_as(
+                    f"Fail to execute command "
+                    f"[echo {clock_event_name} > {self.unbind_clockevent}]."
+                ).is_equal_to(0)
+                timout_timer = create_timer()
+                timeout = 30
+                while timout_timer.elapsed(False) < timeout:
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: It looks this logic is in another test case, use a method to avoid copy/paste code.

Comment (glima) [reply microsoft/testsuites/core/timesync.py]: Or encapsulate it under the already used "retry" module

File: microsoft/testsuites/core/timesync.py [hunk 40]
```diff
@@ -1,9 +1,29 @@
...
+
+@retry(tries=150, delay=0.5)
+def _check_value(
+    node: Node,
+    path: str,
+    expected_value: Optional[Union[str, List[str]]],
+) -> str:
+    cat = node.tools[Cat]
+    result = cat.run(path, force_run=True)
+    if isinstance(expected_value, List):
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: Does it work? the `List` may need to be `list`.

File: microsoft/testsuites/core/timesync.py [hunk 41]
```diff
@@ -1,9 +1,29 @@
...
 
-from lisa import Environment, Node, TestCaseMetadata, TestSuite, TestSuiteMetadata
-from lisa.tools import Cat, Dmesg
+from lisa import Node, TestCaseMetadata, TestSuite, TestSuiteMetadata
+from lisa.tools import Cat, Dmesg, Lscpu
+from lisa.tools.lscpu import CpuType
+
+
+@retry(tries=150, delay=0.5)
+def _check_value(
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: the name can be more accurate like `_check_file_content`.

File: microsoft/testsuites/core/timesync.py [hunk 42]
```diff
@@ -1,9 +1,29 @@
...
+) -> str:
+    cat = node.tools[Cat]
+    result = cat.run(path, force_run=True)
+    if isinstance(expected_value, List):
+        if result.stdout in expected_value:
+            return result.stdout
+    else:
+        if expected_value == result.stdout:
+            return result.stdout
+    return result.stdout
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: The logic looks complex with multiple exit entries. And it looks will return the result, even the expected value is not here. It's same as below logic, is it expected?

```python
result = cat.run(path, force_run=True)
return result.stdout
```

File: microsoft/testsuites/core/timesync.py (hunk 40 above)
Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: The retry works only if there are exception raised.

Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: does return will exist the retry?

File: microsoft/testsuites/core/timesync.py (hunk 42 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: sometimes we need some time to wait for the file content change, if there is not return in this method, we have to cat the file again after this method, then compare actual and expected results.

File: microsoft/testsuites/core/timesync.py (hunk 40 above)
Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: If no exception, it won't retry.

File: microsoft/testsuites/core/timesync.py (hunk 42 above)
Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: The above code paths has no difference. All code paths just return the result.stdout.

File: microsoft/testsuites/core/timesync.py [hunk 43]
```diff
@@ -1,9 +1,33 @@
...
 
-from lisa import Environment, Node, TestCaseMetadata, TestSuite, TestSuiteMetadata
-from lisa.tools import Cat, Dmesg
+from lisa import Node, TestCaseMetadata, TestSuite, TestSuiteMetadata
+from lisa.tools import Cat, Dmesg, Lscpu
+from lisa.tools.lscpu import CpuType
+from lisa.util.perf_timer import create_timer
+
+
+def _check_file_content(
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: how about `_wait_file_changed`

**Review by squirrelsc — DISMISSED** [review]

**Review by squirrelsc — APPROVED** [review]
